File: 05_Binary_Search/05_Search_In_Rotated_Sorted_Array/0033-search-in-rotated-sorted-array.cpp
/*
Problem: LeetCode 33 - Search in Rotated Sorted Array

Description:
Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand. You are given a target value to search for. If found in the array, return its index, otherwise return -1.
You may assume no duplicate exists in the array.

Intuition:
To search for a target value efficiently in the rotated sorted array, we can utilize a modified version of binary search. By comparing the target value with the array elements and adjusting the search space based on the rotation point, we can find the target value or determine its absence.

Approach:
1. Initialize the left and right pointers to the start and end of the array.
2. While the left pointer is less than or equal to the right pointer:
   - Calculate the middle index as (left + right) / 2.
   - If the middle element is equal to the target value, return the middle index.
   - If the left half of the array is sorted:
     - If the target value is within the range of the left half, update the right pointer to middle - 1.
     - Otherwise, update the left pointer to middle + 1.
   - If the right half of the array is sorted:
     - If the target value is within the range of the right half, update the left pointer to middle + 1.
     - Otherwise, update the right pointer to middle - 1.
3. If the target value is not found after the while loop, return -1.

Time Complexity:
The time complexity is O(log n), where n is the size of the array. At each step, the search space is divided in half.

Space Complexity:
The space complexity is O(1), as the algorithm uses a constant amount of extra space.
*/

class Solution {
  public:
    int search(vector<int> &nums, int target) {
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;

            if (nums[middle] == target) { // If middle element is equal to the target value, return the middle index
                return middle;
            }

            if (nums[left] <= nums[middle]) { // If left half of the array is sorted
                if (target >= nums[left] && target < nums[middle]) { // If target value is within the range of the left half
                    right = middle - 1; // Update the right pointer to search in the left half
                } else {
                    left = middle + 1; // Update the left pointer to search in the right half
                }
            } else { // If right half of the array is sorted
                if (target > nums[middle] && target <= nums[right]) { // If target value is within the range of the right half
                    left = middle + 1; // Update the left pointer to search in the right half
                } else {
                    right = middle - 1; // Update the right pointer to search in the left half
                }
            }
        }

        return -1; // Target value not found, return -1
    }
};

File: 05_Binary_Search/05_Search_In_Rotated_Sorted_Array/0033-search-in-rotated-sorted-array.rs
/*
Problem: LeetCode 33 - Search in Rotated Sorted Array

Key Idea:
The key idea is to use modified binary search. Compare the middle element with the left and right endpoints to determine if the left or right half is sorted. Adjust the search range accordingly until you find the target or exhaust the search space.

Approach:
1. Initialize two pointers, `left` and `right`, to the start and end of the array.
2. In each step of the binary search:
   a. Calculate the middle index as `mid` = (left + right) / 2.
   b. Check if `nums[mid]` is equal to the target. If true, return `mid` as the result.
   c. Determine if the left half or right half of the array is sorted by comparing `nums[left]` and `nums[mid]`.
      - If the left half is sorted:
        i. Check if the target falls within the range of `nums[left]` and `nums[mid]`. If true, update `right = mid - 1`.
        ii. Otherwise, update `left = mid + 1`.
      - If the right half is sorted:
        i. Check if the target falls within the range of `nums[mid]` and `nums[right]`. If true, update `left = mid + 1`.
        ii. Otherwise, update `right = mid - 1`.
3. If the target is not found after the binary search, return -1 to indicate that it's not in the array.

Time Complexity:
- The binary search takes O(log n) time, where n is the length of the input array.

Space Complexity:
- The space complexity is O(1) as we are using only a constant amount of extra space.
*/

impl Solution {
    pub fn search(nums: Vec<i32>, target: i32) -> i32 {
        let mut left = 0;
        let mut right = nums.len() - 1;

        while left <= right {
            let mid = left + (right - left) / 2;

            if nums[mid] == target {
                return mid as i32;
            }

            if nums[left] <= nums[mid] {
                if nums[left] <= target && target < nums[mid] {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if nums[mid] < target && target <= nums[right] {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }

        -1
    }
}
